diesel/src/associations/belongs_to.rs

```diff
@@ -28,6 +29,23 @@ pub trait BelongsTo<Parent> {
     fn foreign_key_column() -> Self::ForeignKeyColumn;
 }
 
+/// A trait to abstract which types could be used in a foreign key context.
+///
+/// The use case of this trait is to abstract away wrapper types used for example
+/// to represent if a foreign key could be null.
+///
+/// You will not need to interact with this trait, unless you are writing an
+/// implementation of custom type wrapper that could be used in foreign key position
+pub trait ForeignKey<ST> {
+    /// The (rust) type of the primary key that is referenced by the foreign key
+    type KeyType;
+
+    /// A helper function to extract the foreign key value from the wrapper struct.
+    ///
+    /// Returns `None` if the value is null otherwise `Some(value)`.
+    fn key(&self) -> Option<&Self::KeyType>;
```

Comment:
This signature won't be able to represent composite keys. Should this take `self` by value, and be implemented on a reference instead?